Rework interface definition to cleanly show openAccount. Add HINTS.md…

ricemery opened the pull request:
1) Rework interface definition to cleanly show openAccount. 
2) Add HINTS.md with note about mutable state. 
3) Add Bank.scala with BankAccount trait that should be implemented, and Bank object that should be filled out by the implementer with BankAccount construction.
Refs #214, #220, #137

abo64 commented on exercises/bank-account/HINTS.md:
```diff
@@ -0,0 +1,4 @@
+## Hints
+
+This exercise is testing mutable state that can be accessed saftely from multiple threads. Scala provides a variety of ways to protect 
+mutable state. For developers familiar with Java concurrency, Scala can utilize the Java concurrency support such as the Java synchronized block.
```
Although this might already be some sort of spoiler you could add something like:
```
In Scala there are two ways to achieve mutable state: Use a "var" or a mutable object.
Two common mistakes here are:
- Do not use a "var" that is also a mutable object. One is enough, but not both together.
- Don't expose the "var" or mutable object to the outside world. So make them "private" and change the mutable object into immutable before you return it as a value.
```
One could also imagine a standard format with a separate section named "Common Pitfalls" or so.

ErikSchierboom replied:
The suggested addendum by @abo64 seems like a good one, even though it is a bit of a spoiler.

abo64 commented on exercises/bank-account/src/main/scala/Bank.scala:
```diff
@@ -0,0 +1,13 @@
+  def incrementBalance(increment: Int): Option[Int]
+}
+
+object Bank {
```
Why don't you want to use the companion object anymore?
Same for the file name. AFAIK the file name is always the same as the class/object to be implemented?

ErikSchierboom replied:
Well, there is a `Bank` object being implemented of course, so that sort of makes sense. I usually named the files after the slug name, so for the `BankAccount` exercise, I would use `BankAccount.scala`.

ricemery replied:
I changed the name for the Object from BankAccount to Bank. I didn't like having a BankAccount.openAccount function. It seems like a Bank should be opening accounts not a BankAccount.  I thought about having a Bank.scala and BankAccount.scala. But, then I didn't like having 2 .scala files for the exercise..

ricemery commented:
I renamed Bank.scala to BankAccount.scala . At least now the source file is named to match the test cases file.
I also added the "Common pitfalls" text to the HINTS.

Thanks!

ricemery commented:
Thanks!
